### PR TITLE
Document Message Identity

### PR DESCRIPTION
**Pull request by mikeminutillo**
Connects to #1651

### CODE REVIEW

**Comment by SimonCropp** [comment]
for history read https://github.com/Particular/NServiceBus/issues/1779

**Comment by mikeminutillo** [comment]
Currently [setting the header in a message mutator](https://github.com/ramonsmits/NServiceBus-Samples/blob/master/MessageIdMutator.v5/MyPublisher/MessageIdFromMessageMutator.cs) won't work in V6. 

Why? 

Because the behavior that executes those mutators happens during the `IOutgoingLogicalMessageContext` which contains it's own definition of what the message Id is. This message Id will eventually be forwarded all the way to [here](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Pipeline/Outgoing/OutgoingPhysicalToRoutingConnector.cs#L12) where it overwrites anything placed in the MessageId header before that.

I'll verify and then raise an issue in NServiceBus

File: Snippets/Core/Core_6/MessageIdentity/MessageIdFromMutator.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+            return Task.FromResult(0);
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+    }
+    #endregion
```
Comment (SimonCropp) [inline Snippets/Core/Core_6/MessageIdentity/MessageIdFromMutator.cs]: i would move this region up to above GenerateIdForMessage

File: Snippets/Core/Core_5/MessageIdentity/MessageIdFromMessageMutator.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IBus bus;
+    }
+    #endregion
```
Comment (SimonCropp) [inline Snippets/Core/Core_5/MessageIdentity/MessageIdFromMessageMutator.cs]: i would move this region up to above GenerateIdForMessage

File: Snippets/Core/Core_5/MessageIdentity/MessageIdFromMessageMutator.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+            return message;
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IBus bus;
```
Comment (SimonCropp) [inline Snippets/Core/Core_5/MessageIdentity/MessageIdFromMessageMutator.cs]: most of our doco has the field at the top

File: Snippets/Core/Core_4/MessageIdentity/MessageIdFromMessageMutator.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IBus bus;
+    }
+    #endregion
```
Comment (SimonCropp) [inline Snippets/Core/Core_4/MessageIdentity/MessageIdFromMessageMutator.cs]: i would move this region up to above GenerateIdForMessage

File: Snippets/Core/Core_3/MessageIdentity/MessageIdFromMessageMutator.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IBus bus;
+    }
+    #endregion
```
Comment (SimonCropp) [inline Snippets/Core/Core_3/MessageIdentity/MessageIdFromMessageMutator.cs]: i would move this region up to above GenerateIdForMessage

File: Snippets/Core/Core_3/MessageIdentity/MessageIdFromMessageMutator.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+            return message;
+        }
+
+        string GenerateIdForMessage(object message)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IBus bus;
```
Comment (SimonCropp) [inline Snippets/Core/Core_3/MessageIdentity/MessageIdFromMessageMutator.cs]: move field to top

**Pull request merged**